enhancements/authentication/pod-security-admission-enforcement.md

```diff
@@ -0,0 +1,428 @@
+
+- With the `security.openshift.io/MinimallySufficientPodSecurityStandard` annotation, it can evaluate Namespaces that lack the `pod-security.kubernetes.io/enforce` label but have user-overridden warn or audit labels.
+- With `ValidatedSCCSubjectType`, the controller can classify issues arising from user-based SCC workloads separately.
+  Many of the remaining clusters with violations appear to involve workloads admitted by user SCCs.
+
+### Secure Rollout
+
+The Proposal section indicates that enforcement will be introduced first at the Namespace level and later at the global (cluster-wide) level.
+In addition to adjusting how the `OpenShiftPodSecurityAdmission` `FeatureGate` behaves, administrators need visibility and control throughout this transition.
+A new API is necessary to provide this flexibility.
+
+#### New API
+
+This API offers a gradual way to roll out Pod Security Admission enforcement to clusters.
+It gives users the ability to influence the rollout and see feedback on which Namespaces might violate Pod Security standards.
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PSAEnforcementMode indicates the actual enforcement state of Pod Security Admission
+// in the cluster. Unlike PSATargetMode, which reflects the user’s desired or “target”
+// setting, PSAEnforcementMode describes the effective mode currently active.
+//
+// The modes define a progression from no enforcement, through label-based enforcement
+// to label-based with global config enforcement.
+type PSAEnforcementMode string
+
+const (
+	// EnforcementModePrivileged indicates that no Pod Security restrictions
+	// are effectively applied.
+	// This aligns with a pre-rollout or fully "privileged" cluster state,
+	// where neither enforce labels are set nor the global config enforces "Restricted".
+	EnforcementModePrivileged PSAEnforcementMode = "Privileged"
+
+	// EnfrocementModeLabel indicates that the cluster is enforcing Pod Security
+	// labels at the Namespace level (via the PodSecurityAdmissionLabelSynchronizationController),
+	// but the global kube-apiserver configuration is still "Privileged."
+	EnfrocementModeLabel PSAEnforcementMode = "LabelEnforcement"
+
+	// EnforcmentModeFull indicates that the cluster is enforcing
+	// labels at the Namespace level, and the global configuration has been set
+	// to "Restricted" on the kube-apiserver.
+	// This represents full enforcement, where both Namespace labels and the global config
+	// enforce Pod Security Admission restrictions.
+	EnforcmentModeFull PSAEnforcementMode = "FullEnforcement"
+)
+
+// PSATargetMode reflects the user’s chosen (“target”) enforcement level.
+type PSATargetMode string
+
+const (
```

Comment:
Should there be an option that is reflective of the `LabelEnforcement` enforcement mode that would be specified in the status?

Reply:
We could, but I mentioned it in the [Non-Goals](https://github.com/openshift/enhancements/blob/f8f491154ab3a462566d11cfc15d29099162c2ad/enhancements/authentication/pod-security-admission-enforcement.md#non-goals).

The reason is, that I want to keep the permutations of possible state as minimal as possible.

Reply:
I don't know that this question relates to the non goals you outlined. What target mode(s) might result in the `LabelEnforcement` enforcement mode being populated in the status?

Reply:
It is part of this [non-goal](https://github.com/openshift/enhancements/pull/1747/files#diff-e48dd33b3f59b545b505293e5c81d96a29a2e9cc994f4527b46bde63a2def9d8R56).

I am not eager to maintain too many states. Either you are privileged, restricted or on the way progressing forward through "LabelOnly" to "FullEnforcement".